### PR TITLE
feat: avoid checking the whole project during initial loading

### PR DESCRIPTION
**Pull request by matklad**
bors r+
🤖

### CODE REVIEW

**Comment by matklad** [comment]
Fun fact -- took me a while to debug this, because cargo caches rustc invocations. So it cached the first, buggy implementation, and happy echoed buggy behavior at me and leaving me wondering why my bug fix doesn't work and why my logs don't show up. Guess who added this caching to cargo in the first place?

**Comment by matklad** [comment]
Turns out that the cache proprely accounts for rustc's own mtime (pat on the back, @matklad!) but misses the wrappers specifically: https://github.com/rust-lang/cargo/pull/9348

**Comment by rishadbaniya** [comment]
Just installed a binary of rust analyzer for neovim, made "main.rs" file empty and then opened it, it showed "cargo check failed" and tells me to press enter and then loops ...............

**Comment by matklad** [comment]
@rishadbaniya  please open a separate issuer with specific steps for reproduction, or ask more open-ended support questions in the forum: https://users.rust-lang.org/c/ide/14

File: crates/rust-analyzer/tests/rust-analyzer/main.rs [hunk 1]
```diff
@@ -527,7 +527,7 @@ version = \"0.0.0\"
 #[test]
 fn out_dirs_check() {
     if skip_slow_tests() {
-        return;
+        // return;
```
Comment (jonas-schievink) [inline crates/rust-analyzer/tests/rust-analyzer/main.rs:530]: Looks like you left this in by accident

Comment (matklad) [reply crates/rust-analyzer/tests/rust-analyzer/main.rs:530]: Good catch, fixed https://github.com/rust-analyzer/rust-analyzer/pull/8584.

I wonder if there's a better way to convenently run a single slow test from the editor that uncommenting the thing...

Comment (Veykril) [reply crates/rust-analyzer/tests/rust-analyzer/main.rs:530]: Maybe an env var that RA can set similar to what we do for updating expect tests via command?

Comment (edwin0cheng) [reply crates/rust-analyzer/tests/rust-analyzer/main.rs:530]: The way I did :

```rust
fn out_dirs_check() {
    if skip_slow_tests() {
        // TODO:
        // return;
    }
}
```

Comment (matklad) [reply crates/rust-analyzer/tests/rust-analyzer/main.rs:530]: Cool, didn't thought about that!

**Pull request merged**